src/codec/decoder/mod.rs

```diff
@@ -15,44 +15,82 @@ pub use self::conceal::Conceal;
 pub mod check;
 pub use self::check::Check;
 
+pub mod opened;
+pub use self::opened::Opened;
+
+use std::ptr;
 use std::ffi::CString;
-use std::slice::from_raw_parts;
 use std::ops::{Deref, DerefMut};
 
 use ffi::*;
-use super::{Id, Profile};
-use super::context::Opened;
-use ::{Codec, Error, Discard, Rational};
-use ::media;
+use super::{Id, Context};
+use ::{Codec, Error, Discard, Dictionary};
 
-pub struct Decoder(pub Opened);
+pub struct Decoder(pub Context);
 
 impl Decoder {
-	pub fn video(self) -> Result<Video, Error> {
-		if self.medium() == media::Type::Video {
-			Ok(Video(self))
-		}
-		else {
-			Err(Error::InvalidData)
+	pub fn open(mut self) -> Result<Opened, Error> {
+		unsafe {
+			match avcodec_open2(self.as_mut_ptr(), ptr::null(), ptr::null_mut()) {
+				0 => Ok(Opened(self)),
+				e => Err(Error::from(e))
+			}
 		}
 	}
 
+	pub fn open_as(mut self, codec: &Codec) -> Result<Opened, Error> {
+        if codec.is_decoder() {
+            unsafe {
+                match avcodec_open2(self.as_mut_ptr(), codec.as_ptr(), ptr::null_mut()) {
+                    0 => Ok(Opened(self)),
+                    e => Err(Error::from(e))
+                }
+            }
+        }
+        else {
+            Err(Error::InvalidData)
+        }
+	}
+
+	pub fn open_as_with(mut self, codec: &Codec, options: Dictionary) -> Result<Opened, Error> {
+        if codec.is_decoder() {
+            unsafe {
+                match avcodec_open2(self.as_mut_ptr(), codec.as_ptr(), &mut options.take()) {
+                    0 => Ok(Opened(self)),
+                    e => Err(Error::from(e))
+                }
+            }
+        }
+        else {
+            Err(Error::InvalidData)
+        }
+	}
+
+	pub fn video(self) -> Result<Video, Error> {
+        if let Some(ref codec) = find(self.id()) {
+            self.open_as(codec).and_then(|o| o.video())
```

Comment:
`self` is by value, `mut self` just indicates that you're going to change something in `self`.